docs/adrs/00004-importer-scalability.md

```diff
@@ -0,0 +1,62 @@
+# 00004. Importer Scalability in Trustify
+
+Date: 2025-02-14
+
+## Status
+
+Draft
+
+## Context
+
+This ADR is concerned with our ability to scale the importer processes
+that ingest documents -- SBOM's and advisories -- from various
+sources.
+
+Currently, the importer process iterates over the enabled importer
+definitions in the database sequentially, running each according to
+its schedule. Executing more importer processes results in redundant
+work, each of them doing the same thing, i.e. there is no coordination
+to distribute the load among them.
+
+### Requirements
+
+* The work should be distributed -- no two importer processes should
+  be doing the same work.
+* The solution should be fault-tolerant -- importer instances may come
+  and go -- any work in process by a crashed/killed one must be
+  eventually picked up by another.
+* In short, adding an importer process should increase the ingestion
+  throughput, i.e. the system should be scalable.
+
+## Decision
+
+Since trustify already depends on PostgreSQL -- the import definitions
+persist there -- we can persist the "coordination state" there, too.
+
+Two columns can be added to the `IMPORTER` table:
```

Comment:
Have we considered a separate database for this kind of thing eg. we can have trustify and trustify-importer served up by single pg instance.

Reply:
The importer's only reason for existing is to populate trustify's database. It seems very odd to consider giving it its own just to coordinate replicas.

Reply:
ya ...anything we can do to promote  loose coupling is a good thing - for me an importer app (that only pushes to a single app) is still a different thing but not suggesting we should fully pull out, just was thinking how we might encourage separation of concerns (in code organisation and database). I would put these kind of musings in Alternate section.